--- lighttpd-1.4.25/configure.ac	2009-11-21 15:11:50.000000000 +0000
+++ lighttpd-1.4.24/configure.ac	2009-09-01 18:34:54.000000000 +0000
@@ -5,9 +5,8 @@
 AC_CONFIG_MACRO_DIR([m4])
 
 AC_CANONICAL_TARGET
 
-AM_INIT_AUTOMAKE([-Wall -Wportability -Wno-override -Werror foreign dist-bzip2 tar-ustar silent-rules color-tests])
-AM_SILENT_RULES
+AM_INIT_AUTOMAKE([-Wall -Wportability -Wno-override -Werror foreign dist-bzip2 tar-ustar])
 
 # Checks for programs.
 AC_PROG_CC
