http://trac.lighttpd.net/trac/ticket/475

Index: branches/lighttpd-merge-1.4.x/src/configparser.y
===================================================================
--- branches/lighttpd-merge-1.4.x/src/configparser.y (revision 1061)
+++ branches/lighttpd-merge-1.4.x/src/configparser.y (revision 1062)
@@ -183,4 +183,5 @@
       array_replace(vars, du);
     }
+    B->free(B);
   } else if (NULL != (du = configparser_get_variable(ctx, A))) {
     du = configparser_merge_data(du, B);
@@ -192,13 +193,11 @@
       array_insert_unique(ctx->current->value, du);
     }
+    B->free(B);
   } else {
-    fprintf(stderr, "Undefined config variable in conditional %d %s: %s\n", 
-            ctx->current->context_ndx,
-            ctx->current->key->ptr, A->ptr);
-    ctx->ok = 0;
+    buffer_copy_string_buffer(B->key, A);
+    array_insert_unique(ctx->current->value, B);
   }
   buffer_free(A);
   A = NULL;
-  B->free(B);
   B = NULL;
 }
