diff -ur lighttpd-1.4.8-o/doc/proxy.txt lighttpd-1.4.8/doc/proxy.txt
--- lighttpd-1.4.8-o/doc/proxy.txt	2005-08-10 16:26:16.000000000 -0600
+++ lighttpd-1.4.8/doc/proxy.txt	2006-01-10 01:55:04.000000000 -0700
@@ -67,6 +67,8 @@
   :"host":      is ip of the proxy server
   :"port":      is tcp-port on the "host" used by the proxy
                 server (default: 80)
+  :"fix-redirects":  rewrite redirects from proxied servers to reflect this
+										 server's hostname and port
 
   e.g.: ::
   
diff -ur lighttpd-1.4.8-o/src/array.h lighttpd-1.4.8/src/array.h
--- lighttpd-1.4.8-o/src/array.h	2005-09-23 12:24:18.000000000 -0600
+++ lighttpd-1.4.8/src/array.h	2006-01-10 00:38:55.000000000 -0700
@@ -129,6 +129,7 @@
 		
 	int usage; /* fair-balancing needs the no. of connections active on this host */
 	int last_used_ndx; /* round robin */
+	short fix_redirects;	
 } data_fastcgi;
 
 data_fastcgi *data_fastcgi_init(void);
diff -ur lighttpd-1.4.8-o/src/data_fastcgi.c lighttpd-1.4.8/src/data_fastcgi.c
--- lighttpd-1.4.8-o/src/data_fastcgi.c	2005-08-23 08:36:12.000000000 -0600
+++ lighttpd-1.4.8/src/data_fastcgi.c	2006-01-10 00:38:55.000000000 -0700
@@ -57,6 +57,7 @@
 	ds->host = buffer_init();
 	ds->port = 0;
 	ds->is_disabled = 0;
+	ds->fix_redirects = 0;
 	
 	ds->copy = data_fastcgi_copy;
 	ds->free = data_fastcgi_free;
diff -ur lighttpd-1.4.8-o/src/mod_proxy.c lighttpd-1.4.8/src/mod_proxy.c
--- lighttpd-1.4.8-o/src/mod_proxy.c	2005-11-18 05:29:36.000000000 -0700
+++ lighttpd-1.4.8/src/mod_proxy.c	2006-01-10 00:58:18.000000000 -0700
@@ -277,6 +277,7 @@
 						{ "host",              NULL, T_CONFIG_STRING, T_CONFIG_SCOPE_CONNECTION },      /* 0 */
 						{ "port",              NULL, T_CONFIG_SHORT, T_CONFIG_SCOPE_CONNECTION },       /* 1 */
 						{ "balance",              NULL, T_CONFIG_STRING, T_CONFIG_SCOPE_CONNECTION },   /* 2 */
+						{ "fix-redirects",     NULL, T_CONFIG_SHORT, T_CONFIG_SCOPE_CONNECTION },       /* 3 */
 						{ NULL,                NULL, T_CONFIG_UNSET, T_CONFIG_SCOPE_UNSET }
 					};
 					
@@ -297,6 +298,7 @@
 					
 					pcv[0].destination = df->host;
 					pcv[1].destination = &(df->port);
+					pcv[3].destination = &(df->fix_redirects);
 					
 					if (0 != config_insert_values_internal(srv, da_host->value, pcv)) {
 						return HANDLER_ERROR;
@@ -527,7 +529,7 @@
 }
 
 
-static int proxy_response_parse(server *srv, connection *con, plugin_data *p, buffer *in) {
+static int proxy_response_parse(server *srv, connection *con, plugin_data *p, buffer *in, handler_ctx *hctx) {
 	char *s, *ns;
 	int http_response_status = -1;
 	
@@ -586,7 +588,17 @@
 			break;
 		case 8:
 			if (0 == strncasecmp(key, "Location", key_len)) {
+				buffer *host;
 				con->parsed_response |= HTTP_LOCATION;
+
+				host = buffer_init();
+				buffer_copy_string_len(host, con->request.http_host, strchr(con->request.http_host->ptr, ':') - con->request.http_host->ptr);
+
+				if(hctx->host->fix_redirects && strncmp(value, "http://", 7) == 0 && strncasecmp(value + 7, host->ptr, host->used) == 0 && *(value + 7 + host->used) == ':' && atoi(value + 7 + host->used + 1) == hctx->host->port) {
+					value = strchr(value + 7 + host->used, '/');
+				}
+
+				buffer_free(host);
 			}
 			break;
 		case 10:
@@ -688,7 +700,7 @@
 				log_error_write(srv, __FILE__, __LINE__, "sb", "Header:", hctx->response_header);
 #endif
 				/* parse the response header */
-				proxy_response_parse(srv, con, p, hctx->response_header);
+				proxy_response_parse(srv, con, p, hctx->response_header, hctx);
 					
 				/* enable chunked-transfer-encoding */
 				if (con->request.http_version == HTTP_VERSION_1_1 &&
