--- lighttpd-1.4.56/src/server.c~	2020-11-30 00:32:01.000000000 +0200
+++ lighttpd-1.4.56/src/server.c	2020-11-30 08:29:32.022380561 +0200
@@ -26,7 +26,7 @@
 # define REPO_VERSION ""
 #endif
 
-#define PACKAGE_DESC PACKAGE_NAME "/" PACKAGE_VERSION REPO_VERSION
+#define PACKAGE_DESC PACKAGE_NAME "/" PACKAGE_VERSION REPO_VERSION " (PLD Linux)"
 static const buffer default_server_tag =
   { PACKAGE_DESC "\0server", sizeof(PACKAGE_DESC), 0 };
 

