--- lighttpd-1.2.2/src/network_backends.h	2004-06-15 17:21:34.000000000 +0000
+++ lighttpd-1.2.2.new/src/network_backends.h	2004-06-15 20:35:32.767494640 +0000
@@ -5,7 +5,7 @@
 
 /* on linux 2.4.x you get either sendfile or LFS */
 #if defined HAVE_SYS_SENDFILE_H && defined HAVE_SENDFILE && (!defined _LARGEFILE_SOURCE || defined HAVE_SENDFILE64) && defined HAVE_WRITEV && defined(__linux__) && !defined HAVE_SENDFILE_BROKEN
-# define USE_LINUX_SENDFILE
+/* # define USE_LINUX_SENDFILE */
 # include <sys/sendfile.h>
 # include <sys/uio.h>
 #endif
