diff -dur lighttpd-1.4.31.orig/configure.ac lighttpd-1.4.31/configure.ac
--- lighttpd-1.4.31.orig/configure.ac	2012-05-31 16:49:18.000000000 +0200
+++ lighttpd-1.4.31/configure.ac	2012-06-08 09:54:13.000000000 +0200
@@ -38,9 +38,6 @@
 
 AC_EXEEXT
 
-dnl more automake stuff
-AM_C_PROTOTYPES
-
 dnl libtool
 AC_DISABLE_STATIC
 AC_ENABLE_SHARED
@@ -23,6 +23,7 @@
 dnl AC_PROG_CXX
 AC_PROG_LN_S
 AC_PROG_MAKE_SET
+AM_PROG_AR
 
 dnl check environment
 AC_AIX
