--- src/network_backends.h~	2004-06-23 20:18:51.000000000 +0000
+++ src/network_backends.h	2004-09-05 11:48:59.869628504 +0000
@@ -5,7 +5,9 @@
 
 /* on linux 2.4.x you get either sendfile or LFS */
 #if defined HAVE_SYS_SENDFILE_H && defined HAVE_SENDFILE && (!defined _LARGEFILE_SOURCE || defined HAVE_SENDFILE64) && defined HAVE_WRITEV && defined(__linux__) && !defined HAVE_SENDFILE_BROKEN
+/* it is brokem on amd64, jan is working to fix it
 # define USE_LINUX_SENDFILE
+*/
 # include <sys/sendfile.h>
 # include <sys/uio.h>
 #endif
