diff -ur lighttpd-1.4.10/src/server.c lighttpd-1.4.10.new/src/server.c
--- lighttpd-1.4.10/src/server.c	2006-02-01 11:50:02.000000000 +0000
+++ lighttpd-1.4.10.new/src/server.c	2006-02-19 09:56:58.000000000 +0000
@@ -218,7 +218,9 @@
 	CLEAN(srvconf.event_handler);
 	CLEAN(srvconf.pid_file);
 	CLEAN(srvconf.modules_dir);
-	
+	CLEAN(srvconf.network_backend);
+	CLEAN(srvconf.upload_tempdirs);	
+
 	CLEAN(tmp_chunk_len);
 #undef CLEAN
 
