Index: configure.in
===================================================================
Index: src/mod_simple_vhost.c
===================================================================
--- src/mod_simple_vhost.c	(.../tags/lighttpd-1.4.21)	(revision 2399)
+++ src/mod_simple_vhost.c	(.../branches/lighttpd-1.4.x)	(revision 2399)
@@ -249,6 +249,8 @@
 				return HANDLER_GO_ON;
 			} else {
 				buffer_copy_string_buffer(con->server_name, p->conf.default_host);
+				buffer_copy_string_buffer(con->physical.doc_root, p->doc_root);
+
 				/* do not cache default host */
 				return HANDLER_GO_ON;
 			}
Index: src/mod_cml_lua.c
===================================================================
--- src/mod_cml_lua.c	(.../tags/lighttpd-1.4.21)	(revision 2399)
+++ src/mod_cml_lua.c	(.../branches/lighttpd-1.4.x)	(revision 2399)
@@ -282,11 +282,11 @@
 
 	/* 2 default constants */
 	lua_pushliteral(L, "CACHE_HIT");
-	lua_pushboolean(L, 0);
+	lua_pushnumber(L, 0);
 	lua_settable(L, LUA_GLOBALSINDEX);
 
 	lua_pushliteral(L, "CACHE_MISS");
-	lua_pushboolean(L, 1);
+	lua_pushnumber(L, 1);
 	lua_settable(L, LUA_GLOBALSINDEX);
 
 	/* load lua program */
Index: SConstruct
===================================================================
Index: NEWS
===================================================================
--- NEWS	(.../tags/lighttpd-1.4.21)	(revision 2399)
+++ NEWS	(.../branches/lighttpd-1.4.x)	(revision 2399)
@@ -3,8 +3,12 @@
 NEWS
 ====
 
-- 1.4.21 -
+- 1.4.22 -
+  * Fix wrong lua type for CACHE_MISS/CACHE_HIT in mod_cml (fixes #533)
+  * Fix default vhost in mod_simple_vhost (fixes #1905)
 
+- 1.4.21 - 2009-02-16
+
   * Fix base64 decoding in mod_auth (#1757, thx guido)
   * Fix mod_cgi segfault when bound to unix domain socket (#653)
   * Do not rely on ioctl FIONREAD (#673)
Index: CMakeLists.txt
===================================================================

Property changes on: .
___________________________________________________________________
Modified: bzr:revision-info
   - timestamp: 2009-02-16 14:42:07.753999949 +0100
committer: Stefan Bühler <stbuehler@web.de>
properties: 
	branch-nick: lighttpd-1.4.x

   + timestamp: 2009-02-17 23:44:47.233000040 +0100
committer: Stefan Bühler <stbuehler@web.de>
properties: 
	branch-nick: lighttpd-1.4.x

Modified: bzr:revision-id:v3-trunk0
   - 1127 stbuehler@web.de-20080728081644-j4cxnhduw8kbt8um
1128 stbuehler@web.de-20080728084246-axvxdtjsrratxixs
1129 stbuehler@web.de-20080729211700-s8v6nq2cu06qesls
1130 stbuehler@web.de-20080729211726-4yxb6e5dva1cn0lz
1131 stbuehler@web.de-20080729211750-4ulzigswx17uciyu
1132 stbuehler@web.de-20080729211850-nliz3kd0m576ztuu
1133 stbuehler@web.de-20080730163440-dg2y2sbf0u4grmn4
1134 stbuehler@web.de-20080730173952-kiutzg6geqy7mick
1135 stbuehler@web.de-20080730193616-9kc2ms7rrhv1lkn7
1136 stbuehler@web.de-20080730211457-z4a6uth1y29glbqh
1137 stbuehler@web.de-20080730213517-b6sjcrdwbmipl334
1138 stbuehler@web.de-20080731102617-2xw8unjfqic7lsew
1139 stbuehler@web.de-20080731102703-q4tu5a6em9y8xdg0
1140 stbuehler@web.de-20080731102729-l6vn5b05w9swqbg5
1141 stbuehler@web.de-20080731102756-oj3d4tnk0l90mj77
1142 stbuehler@web.de-20080731204442-blw14cj2fkr3l8ly
1143 stbuehler@web.de-20080731204508-imtfnurf922mg7tj
1144 stbuehler@web.de-20080801112347-girnwswdkwm8wuip
1145 stbuehler@web.de-20080801161245-kx1temr529o7xko9
1146 stbuehler@web.de-20080801175332-oc9e7x8edn1owcc0
1147 stbuehler@web.de-20080801183454-5i66v0gsdv0cgmia
1148 stbuehler@web.de-20080801192849-6zklfbb832sx0hvr
1149 stbuehler@web.de-20080801203119-o16elp8w854s6lol
1150 stbuehler@web.de-20080802162146-a4v57svc788pwdsv
1151 stbuehler@web.de-20080802162202-9udlc1wuwt09pyh2
1152 stbuehler@web.de-20080804135803-yuor9ze06px7qta4
1153 stbuehler@web.de-20080812194728-fupt781o6q058unh
1154 stbuehler@web.de-20080818162116-piz0ukqsaecv2li2
1155 stbuehler@web.de-20080818235700-94t0xc6ml70zojwq
1156 stbuehler@web.de-20080819163650-1qhwsqszr78cr4xx
1157 stbuehler@web.de-20080819163757-1qq3t1f1wj69t8xs
1158 stbuehler@web.de-20080819163914-rklhkurg8apv85l2
1159 stbuehler@web.de-20080819163953-tlqew751e43phf5b
1160 stbuehler@web.de-20080819164108-8ogh68sm1uyteawe
1161 stbuehler@web.de-20080819173911-w5bqpb7cp9jmdqye
1162 stbuehler@web.de-20080819222242-c0ta5gnli9p3j35a
1163 stbuehler@web.de-20080820100730-g1bwdh4nqb53ag9u
1164 stbuehler@web.de-20080820100752-9pggugdyfnnps8qu
1165 stbuehler@web.de-20080820164258-v2j00motsrsc5esp
1166 stbuehler@web.de-20080827144628-hi9hf4ch3n1wf9ao
1167 stbuehler@web.de-20080827144903-tfxu4yehlyu5kegc
1168 stbuehler@web.de-20080827155155-7mt92orehbxkh2lh
1169 stbuehler@web.de-20080917142048-zbcwpk39q9ewd516
1170 stbuehler@web.de-20080917142300-16gzt21x4nbjtj87
1171 stbuehler@web.de-20080919160134-385anjnd3txxdw3x
1172 stbuehler@web.de-20080920134142-fvvwaw2ys51dg4rj
1173 stbuehler@web.de-20080921153311-1f7rn01atdilmxmy
1174 stbuehler@web.de-20080922101346-wel327kjmykkpvmp
1175 stbuehler@web.de-20080923190422-uow06l38ndue36o4
1176 stbuehler@web.de-20080930112012-53jby2m8xslmd1hm
1177 stbuehler@web.de-20080930134824-j9q72rwuiczzof5k
1178 stbuehler@web.de-20080930142037-32pb6m3zjcwryw1w
1179 stbuehler@web.de-20080930142756-ueovgoshyb996bce
1180 stbuehler@web.de-20080930152935-1zfy67brol3xdbc0
1181 stbuehler@web.de-20080930193919-13n2q4c6fbgw0dkx
1182 stbuehler@web.de-20080930211152-4hmgs95wyg2deol7
1183 stbuehler@web.de-20081001132402-hxnyu6yzyk3mjf4d
1184 stbuehler@web.de-20081001155102-qf0mmu2kkpgr7xf0
1185 stbuehler@web.de-20081001160009-n67ss0vzlac2y60k
1186 stbuehler@web.de-20081001200802-l5og517etnneitk0
1188 stbuehler@web.de-20081004160711-ygaohrecmutiqlla
1189 stbuehler@web.de-20081004211932-vq16u26mthbeed7d
1191 stbuehler@web.de-20081005224446-1bztt6zqrjh8w8fd
1192 stbuehler@web.de-20081012114652-ihgz590f0gl5gxpw
1193 stbuehler@web.de-20081012114716-jnzljhexi4z2gh92
1195 stbuehler@web.de-20081016120614-kz39vxtz1pebho0o
1196 stbuehler@web.de-20081016121103-trug4hts0o62d1ut
1197 stbuehler@web.de-20081016121114-65quosenmso8frf8
1198 stbuehler@web.de-20081016121421-xjjb7fb53pxu6odj
1199 stbuehler@web.de-20081205222033-6qok7y19pwp3kxm9
1200 stbuehler@web.de-20081205222811-49izmzxui0y9ncq6
1201 stbuehler@web.de-20081205233903-708beaujtf26gprx
1202 stbuehler@web.de-20081207151631-yv9bdf94zw83jxpv
1203 stbuehler@web.de-20081207151822-mhyg0gkedmttdqvd
1204 stbuehler@web.de-20081207151835-1m3yta2fjc4pgb8y
1205 stbuehler@web.de-20081218221139-w8los43bjbhy9urh
1206 stbuehler@web.de-20081218222305-5wz7000a62iqa81r
1208 stbuehler@web.de-20090203201352-ivan8lsb3nkv1go5
1209 stbuehler@web.de-20090203204231-03zjmk7qiol9yxgq
1210 stbuehler@web.de-20090203210157-bx1e59fqple5oj3v
1211 stbuehler@web.de-20090203221006-qd6w80m7lmeqgrjh
1212 stbuehler@web.de-20090203225303-3dwmialad2u720h8
1213 stbuehler@web.de-20090204102521-jl3vo2ftp5rsbx9y
1214 stbuehler@web.de-20090204151616-n56of74dydkqdkgh
1215 stbuehler@web.de-20090204172956-6wzsv0nm5nxcgfym
1216 stbuehler@web.de-20090205105134-6i5key9439wspueq
1217 stbuehler@web.de-20090205114017-0voscqjd5bdm9mwv
1218 stbuehler@web.de-20090205114442-peekxwpevjl3t7j3
1219 stbuehler@web.de-20090205215425-vicbc6hzb3at7gj9
1220 stbuehler@web.de-20090205220741-vqz9l1eui3dwnulq
1221 stbuehler@web.de-20090205222705-8179v6jkv2x38l70
1222 stbuehler@web.de-20090210194631-6epujtpen9xfxx5j
1223 stbuehler@web.de-20090216134207-fg99ikt1ds21hx25

   + 1127 stbuehler@web.de-20080728081644-j4cxnhduw8kbt8um
1128 stbuehler@web.de-20080728084246-axvxdtjsrratxixs
1129 stbuehler@web.de-20080729211700-s8v6nq2cu06qesls
1130 stbuehler@web.de-20080729211726-4yxb6e5dva1cn0lz
1131 stbuehler@web.de-20080729211750-4ulzigswx17uciyu
1132 stbuehler@web.de-20080729211850-nliz3kd0m576ztuu
1133 stbuehler@web.de-20080730163440-dg2y2sbf0u4grmn4
1134 stbuehler@web.de-20080730173952-kiutzg6geqy7mick
1135 stbuehler@web.de-20080730193616-9kc2ms7rrhv1lkn7
1136 stbuehler@web.de-20080730211457-z4a6uth1y29glbqh
1137 stbuehler@web.de-20080730213517-b6sjcrdwbmipl334
1138 stbuehler@web.de-20080731102617-2xw8unjfqic7lsew
1139 stbuehler@web.de-20080731102703-q4tu5a6em9y8xdg0
1140 stbuehler@web.de-20080731102729-l6vn5b05w9swqbg5
1141 stbuehler@web.de-20080731102756-oj3d4tnk0l90mj77
1142 stbuehler@web.de-20080731204442-blw14cj2fkr3l8ly
1143 stbuehler@web.de-20080731204508-imtfnurf922mg7tj
1144 stbuehler@web.de-20080801112347-girnwswdkwm8wuip
1145 stbuehler@web.de-20080801161245-kx1temr529o7xko9
1146 stbuehler@web.de-20080801175332-oc9e7x8edn1owcc0
1147 stbuehler@web.de-20080801183454-5i66v0gsdv0cgmia
1148 stbuehler@web.de-20080801192849-6zklfbb832sx0hvr
1149 stbuehler@web.de-20080801203119-o16elp8w854s6lol
1150 stbuehler@web.de-20080802162146-a4v57svc788pwdsv
1151 stbuehler@web.de-20080802162202-9udlc1wuwt09pyh2
1152 stbuehler@web.de-20080804135803-yuor9ze06px7qta4
1153 stbuehler@web.de-20080812194728-fupt781o6q058unh
1154 stbuehler@web.de-20080818162116-piz0ukqsaecv2li2
1155 stbuehler@web.de-20080818235700-94t0xc6ml70zojwq
1156 stbuehler@web.de-20080819163650-1qhwsqszr78cr4xx
1157 stbuehler@web.de-20080819163757-1qq3t1f1wj69t8xs
1158 stbuehler@web.de-20080819163914-rklhkurg8apv85l2
1159 stbuehler@web.de-20080819163953-tlqew751e43phf5b
1160 stbuehler@web.de-20080819164108-8ogh68sm1uyteawe
1161 stbuehler@web.de-20080819173911-w5bqpb7cp9jmdqye
1162 stbuehler@web.de-20080819222242-c0ta5gnli9p3j35a
1163 stbuehler@web.de-20080820100730-g1bwdh4nqb53ag9u
1164 stbuehler@web.de-20080820100752-9pggugdyfnnps8qu
1165 stbuehler@web.de-20080820164258-v2j00motsrsc5esp
1166 stbuehler@web.de-20080827144628-hi9hf4ch3n1wf9ao
1167 stbuehler@web.de-20080827144903-tfxu4yehlyu5kegc
1168 stbuehler@web.de-20080827155155-7mt92orehbxkh2lh
1169 stbuehler@web.de-20080917142048-zbcwpk39q9ewd516
1170 stbuehler@web.de-20080917142300-16gzt21x4nbjtj87
1171 stbuehler@web.de-20080919160134-385anjnd3txxdw3x
1172 stbuehler@web.de-20080920134142-fvvwaw2ys51dg4rj
1173 stbuehler@web.de-20080921153311-1f7rn01atdilmxmy
1174 stbuehler@web.de-20080922101346-wel327kjmykkpvmp
1175 stbuehler@web.de-20080923190422-uow06l38ndue36o4
1176 stbuehler@web.de-20080930112012-53jby2m8xslmd1hm
1177 stbuehler@web.de-20080930134824-j9q72rwuiczzof5k
1178 stbuehler@web.de-20080930142037-32pb6m3zjcwryw1w
1179 stbuehler@web.de-20080930142756-ueovgoshyb996bce
1180 stbuehler@web.de-20080930152935-1zfy67brol3xdbc0
1181 stbuehler@web.de-20080930193919-13n2q4c6fbgw0dkx
1182 stbuehler@web.de-20080930211152-4hmgs95wyg2deol7
1183 stbuehler@web.de-20081001132402-hxnyu6yzyk3mjf4d
1184 stbuehler@web.de-20081001155102-qf0mmu2kkpgr7xf0
1185 stbuehler@web.de-20081001160009-n67ss0vzlac2y60k
1186 stbuehler@web.de-20081001200802-l5og517etnneitk0
1188 stbuehler@web.de-20081004160711-ygaohrecmutiqlla
1189 stbuehler@web.de-20081004211932-vq16u26mthbeed7d
1191 stbuehler@web.de-20081005224446-1bztt6zqrjh8w8fd
1192 stbuehler@web.de-20081012114652-ihgz590f0gl5gxpw
1193 stbuehler@web.de-20081012114716-jnzljhexi4z2gh92
1195 stbuehler@web.de-20081016120614-kz39vxtz1pebho0o
1196 stbuehler@web.de-20081016121103-trug4hts0o62d1ut
1197 stbuehler@web.de-20081016121114-65quosenmso8frf8
1198 stbuehler@web.de-20081016121421-xjjb7fb53pxu6odj
1199 stbuehler@web.de-20081205222033-6qok7y19pwp3kxm9
1200 stbuehler@web.de-20081205222811-49izmzxui0y9ncq6
1201 stbuehler@web.de-20081205233903-708beaujtf26gprx
1202 stbuehler@web.de-20081207151631-yv9bdf94zw83jxpv
1203 stbuehler@web.de-20081207151822-mhyg0gkedmttdqvd
1204 stbuehler@web.de-20081207151835-1m3yta2fjc4pgb8y
1205 stbuehler@web.de-20081218221139-w8los43bjbhy9urh
1206 stbuehler@web.de-20081218222305-5wz7000a62iqa81r
1208 stbuehler@web.de-20090203201352-ivan8lsb3nkv1go5
1209 stbuehler@web.de-20090203204231-03zjmk7qiol9yxgq
1210 stbuehler@web.de-20090203210157-bx1e59fqple5oj3v
1211 stbuehler@web.de-20090203221006-qd6w80m7lmeqgrjh
1212 stbuehler@web.de-20090203225303-3dwmialad2u720h8
1213 stbuehler@web.de-20090204102521-jl3vo2ftp5rsbx9y
1214 stbuehler@web.de-20090204151616-n56of74dydkqdkgh
1215 stbuehler@web.de-20090204172956-6wzsv0nm5nxcgfym
1216 stbuehler@web.de-20090205105134-6i5key9439wspueq
1217 stbuehler@web.de-20090205114017-0voscqjd5bdm9mwv
1218 stbuehler@web.de-20090205114442-peekxwpevjl3t7j3
1219 stbuehler@web.de-20090205215425-vicbc6hzb3at7gj9
1220 stbuehler@web.de-20090205220741-vqz9l1eui3dwnulq
1221 stbuehler@web.de-20090205222705-8179v6jkv2x38l70
1222 stbuehler@web.de-20090210194631-6epujtpen9xfxx5j
1223 stbuehler@web.de-20090216134207-fg99ikt1ds21hx25
1224 stbuehler@web.de-20090217085833-9g5c7j7zdylvezl5
1225 stbuehler@web.de-20090217133414-y80hydn9raqgkgto
1226 stbuehler@web.de-20090217224447-ve7ns45c9otbgz9h


