--- src/mod_auth.c	(revision 10)
+++ src/mod_auth.c	(revision 11)
@@ -404,17 +404,17 @@
 			}
 					
 			if (method == NULL) {
-				log_error_write(srv, __FILE__, __LINE__, "sssbs", "missing entry for key: ", "auth.require", "[", "method", "](string)");
+				log_error_write(srv, __FILE__, __LINE__, "sssss", "missing entry for key: ", "auth.require", "[", "method", "](string)");
 				return HANDLER_ERROR;
 			}
 			
 			if (realm == NULL) {
-				log_error_write(srv, __FILE__, __LINE__, "sssbs", "missing entry for key: ", "auth.require", "[", "realm", "](string)");
+				log_error_write(srv, __FILE__, __LINE__, "sssss", "missing entry for key: ", "auth.require", "[", "realm", "](string)");
 				return HANDLER_ERROR;
 			}
 			
 			if (require == NULL) {
-				log_error_write(srv, __FILE__, __LINE__, "sssbs", "missing entry for key: ", "auth.require", "[", "require", "](string)");
+				log_error_write(srv, __FILE__, __LINE__, "sssss", "missing entry for key: ", "auth.require", "[", "require", "](string)");
 				return HANDLER_ERROR;
 			}
 			
