--- configure.in-orig	2004-04-16 21:24:11.000000000 +0000
+++ configure.in	2004-04-16 21:25:04.000000000 +0000
@@ -286,7 +286,7 @@
 fi
 
 if test "${GCC}" = "yes"; then
-       CFLAGS="${CFLAGS} -Wall -W -O3 -Wshadow -pedantic"
+       CFLAGS="${CFLAGS} -Wall -W -Wshadow -pedantic"
        if test "${USER}" = "weigon"; then
        	  CFLAGS="${CFLAGS} -march=athlon-mp -fexpensive-optimizations -pipe"
 	  MAKE="${MAKE} -j4"
