--- src/mod_fastcgi.c~	2005-02-20 16:59:47.000000000 +0200
+++ src/mod_fastcgi.c	2005-02-27 14:09:29.000000000 +0200
@@ -1424,7 +1424,7 @@
 	} else {
 		fcgi_addr_in.sin_family = AF_INET;
 		if (INADDR_NONE == (fcgi_addr_in.sin_addr.s_addr = inet_addr(host->host->ptr))) {
-			log_error_write(srv, __FILE__, __LINE__, "sb", 
+			log_error_write(srv, __FILE__, __LINE__, "sbs", 
 					"converting IP-adress failed for", host->host, 
 					"\nBe sure to specify an IP address here");
 			
