Index: /branches/lighttpd-1.4.x/src/connections.c
===================================================================
--- /branches/lighttpd-1.4.x/src/connections.c (revision 1716)
+++ /branches/lighttpd-1.4.x/src/connections.c (revision 1852)
@@ -1433,4 +1433,5 @@
 					/* error-handler is back and has generated content */
 					/* if Status: was set, take it otherwise use 200 */
+					con->http_status = con->error_handler_saved_status;
 				}
 
