; http://redmine.lighttpd.net/issues/2486
diff -urN lighttpd-1.4.32.org/src/http_auth.c lighttpd-1.4.32/src/http_auth.c
--- lighttpd-1.4.32.org/src/http_auth.c	2012-05-17 11:29:24.000000000 +0200
+++ lighttpd-1.4.32/src/http_auth.c	2013-06-05 19:58:07.828450230 +0200
@@ -688,6 +688,8 @@
 		salt[salt_len] = '\0';
 
 		crypted = crypt(pw, salt);
+		if (NULL == crypted)
+			return -1;
 
 		if (0 == strcmp(password->ptr, crypted)) {
 			return 0;
