--- lighttpd-1.4.20/src/mod_proxy.c~	2009-01-16 19:00:15.000000000 +0200
+++ lighttpd-1.4.20/src/mod_proxy.c	2009-01-16 19:06:16.130587778 +0200
@@ -454,6 +454,7 @@
 
 		if (ds->value->used && ds->key->used) {
 			if (buffer_is_equal_string(ds->key, CONST_STR_LEN("Connection"))) continue;
+			if (buffer_is_equal_string(ds->key, CONST_STR_LEN("Proxy-Connection"))) continue;
 
 			buffer_append_string_buffer(b, ds->key);
 			buffer_append_string_len(b, CONST_STR_LEN(": "));
