
TODO: upstream it:
https://github.com/lighttpd/lighttpd1.4/pull/69

--- lighttpd-1.4.56/tests/LightyTest.pm~	2020-11-30 00:32:01.000000000 +0200
+++ lighttpd-1.4.56/tests/LightyTest.pm	2020-11-30 08:31:44.175860280 +0200
@@ -81,7 +81,7 @@
 	if (exists $ENV{LIGHTTPD_EXE_PATH}) {
 		$self->{LIGHTTPD_PATH} = $ENV{LIGHTTPD_EXE_PATH};
 	}
-	$self->{PORT} = 2048;
+	$self->{PORT} = $ENV{LIGHTTPD_TEST_PORT} ? int($ENV{LIGHTTPD_TEST_PORT}) : 2048;
 
 	my ($name, $aliases, $addrtype, $net) = gethostbyaddr(inet_aton("127.0.0.1"), AF_INET);
 
